### ISSUE TITLE
ruffle executes wrong frame scripts on movieclips when registered within a frame script

### ISSUE THREAD

**Issue by nivkner** [ruffle-rs/ruffle#14629]
### Describe the bug

it seems like when adding a child with its own children via `addChild` while in a frame, as opposed to it being part of the frame,
the frame scripts of the grandchildren can execute before the child finished its construction.

i made an SWF (attached below) based on the `button_nested_frame_simple` test which demonstrates this behavior.

when running the SWF with ruffle with the command line:
```
RUST_LOG=warn,ruffle_core=debug,avm_trace=trace ./ruffle_desktop -g gl test.swf
```

i get the following output:

```
2024-01-06T00:27:30.436544Z  WARN egui_wgpu::renderer: Detected a linear (sRGBA aware) framebuffer Rgba8UnormSrgb. egui prefers Rgba8Unorm or Bgra8Unorm    
2024-01-06T00:27:30.481086Z  INFO ruffle_core::library: Loaded new device font "Noto Sans" from swf tag
2024-01-06T00:27:30.495783Z  INFO ruffle_core::player: Loaded SWF version 16, resolution 550x400 @ 24 FPS
2024-01-06T00:27:30.647856Z  INFO run_frame:run_all_phases_avm2: avm_trace: Main Constructor
2024-01-06T00:27:30.647967Z  INFO run_frame:run_all_phases_avm2: avm_trace: Main addFrameScript
2024-01-06T00:27:30.647991Z  INFO run_frame:run_all_phases_avm2: avm_trace: Main frame1
2024-01-06T00:27:30.650296Z  INFO run_frame:run_all_phases_avm2: avm_trace: Main frame2
2024-01-06T00:27:30.650380Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyContainer Constructor
2024-01-06T00:27:30.650403Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild Constructor
2024-01-06T00:27:30.650416Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild addFrameScript
2024-01-06T00:27:30.650430Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild Constructor
2024-01-06T00:27:30.650442Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild addFrameScript
2024-01-06T00:27:30.650622Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyContainer addFrameScript
2024-01-06T00:27:30.650649Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyContainer frame1
2024-01-06T00:27:30.650657Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild frame1
2024-01-06T00:27:30.650663Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild frame1
2024-01-06T00:27:30.671357Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild frame2
2024-01-06T00:27:30.671384Z  INFO run_frame:run_all_phases_avm2: avm_trace: MyChild frame2
```

### Expected behavior

expected to get the same output as the flash player.

when running the same SWF with flash player (debug version), the contents of the log are:
```
Main Constructor
Main addFrameScript
Main frame1
Main frame2
MyContainer Constructor
MyChild Constructor
MyChild addFrameScript
MyChild Constructor
MyChild addFrameScript
MyChild frame1
MyChild frame1
MyContainer addFrameScript
MyChild frame2
MyChild frame2
```

### Content Location

here is the zip containing the SWF as `test.swf`, along with the scripts and `fla` file.

[test.zip](https://github.com/ruffle-rs/ruffle/files/13847717/test.zip)

### Affected platform

Desktop app

### Operating system

Fedora Linux 39.20240104.0 (Silverblue)

### Additional information
tested on:
```
ruffle 0a6a95e6f1bb9becd6dc97e590a06f90a75c083c
rustc 1.77.0-nightly (f688dd684 2024-01-04)
```

**Comment by Lord-McSweeney** [comment]
@Aaron1011 is this a known issue?

**Comment by Aaron1011** [comment]
Yes - I thought I had opened an issue for it, but I guess not

**Comment by nivkner** [comment]
is it a duplicate of #12443? i just noticed it

**Comment by n0samu** [comment]
Was about to ask the same

**Comment by Aaron1011** [comment]
Buttons are particularly awful, but there's also a more general issue with changes to displaylists in the middle of framescript execution. I think we might need a separate passes that queues framscripts for execution in a single list

**Comment by nivkner** [comment]
been a while, but I have decided to try and tackle this issue, which i think is connected to #10443 (that is in fact what motivated this issue, i just thought a simplifed test case would be more useful).

before i write any code, i figured it would be a good idea to document what i learnt, in case anyone else has a solution.

since the handling of buttons in particular is addressed in the aforementioned issue, and for the sake of simplification, i removed the buttons from the attached flash file, and as of 2025-03-26, the behavior of the frame-scripts is still wrong.

on flash player:
```
Main Constructor
Main addFrameScript
Main frame1
Main frame2
MyContainer Constructor
MyChild Constructor
MyChild addFrameScript
MyChild Constructor
MyChild addFrameScript
MyContainer addFrameScript
MyChild frame1
MyChild frame1
MyContainer frame1
MyChild frame2
MyChild frame2
```

on ruffle:
```
2025-03-26T18:51:43.364576Z  INFO avm_trace: Main Constructor
2025-03-26T18:51:43.364638Z  INFO avm_trace: Main addFrameScript
2025-03-26T18:51:43.364692Z  INFO avm_trace: Main frame1
2025-03-26T18:51:43.370737Z  INFO avm_trace: Main frame2
2025-03-26T18:51:43.370803Z  INFO avm_trace: MyContainer Constructor
2025-03-26T18:51:43.370818Z  INFO avm_trace: MyChild Constructor
2025-03-26T18:51:43.370830Z  INFO avm_trace: MyChild addFrameScript
2025-03-26T18:51:43.370843Z  INFO avm_trace: MyChild Constructor
2025-03-26T18:51:43.370847Z  INFO avm_trace: MyChild addFrameScript
2025-03-26T18:51:43.370853Z  INFO avm_trace: MyContainer addFrameScript
2025-03-26T18:51:43.370881Z  INFO avm_trace: MyContainer frame1
2025-03-26T18:51:43.370886Z  INFO avm_trace: MyChild frame1
2025-03-26T18:51:43.370888Z  INFO avm_trace: MyChild frame1
2025-03-26T18:51:43.370922Z  INFO avm_trace: MyChild frame2
2025-03-26T18:51:43.370925Z  INFO avm_trace: MyChild frame2
```

i managed to isolate the cause of the discrepancy to the **calls to `addFrameScripts` during the frame-script phase**.
that is to say, this is not due to objects being dynamically constructed,
since i produced the same order with objects created statically using the `placeObject` tag,
with the call to to `addFrameScript` done separately.

specifically, i believe the behavior of frame script execution is as follows:
at the start of the frame-script phase, scripts that were already registered are executed by traversing the object tree, in order,
executing each object's script.
if during the script phase, a new script is registered **for the current frame**, it is enqueued onto a global queue.
once the traversal is done, the scripts in the queue are executed in order, independently of their position in the tree.

scripts registered for a different frame, including the same frame on a new cycle, are executed in the usual order.
there is a difference on the next cycle when the root frame-script is executed, but that is already handled correctly by ruffle,
and so doesn't seem relevant.

calls to `gotoAndPlay` seem to discard the queue, causing the scripts to execute in the tree order, as if we started a new frame, even if you go to the same frame.

consider for example, the following object tree:
```
       B -- C
      /
R -- A 
      \
       D
```
and suppose the frame-scripts of B are registered at the construction phase, while those of C, A and D are registered at the frame-script of R, in that order.

what we get is that in the frame where the registrations happened, the execution order is B -> C -> A -> D, while in the rest of the frames, the order is A -> B -> C -> D, same as the construction order.

to confirm this, the attached flash files below contain an exhaustive combination of registration orders for the above tree:
[framescript_tests.zip](https://github.com/user-attachments/files/19474697/framescript_tests.zip)
the results are consistent with the above interpretation.

**Comment by nivkner** [comment]
EDIT: will move the remaining test cases to a separate issue
